parsec/api/data/invite.py

```diff
@@ -0,0 +1,119 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPLv3 2019 Scille SAS
+
+from typing import Optional, Tuple, List
+from random import randint, shuffle
+
+from parsec.crypto import VerifyKey, PublicKey, PrivateKey, SecretKey
+from parsec.serde import fields, post_load
+from parsec.api.protocol import (
+    DeviceID,
+    DeviceIDField,
+    DeviceName,
+    DeviceNameField,
+    HumanHandle,
+    HumanHandleField,
+)
+from parsec.api.data.base import BaseAPIData, BaseSchema
+
+
+def generate_sas_codes(
+    claimer_nonce: bytes, greeter_nonce: bytes, shared_secret_key: SecretKey
+) -> Tuple[int, int]:
+    # Computes combined HMAC
+    combined_nonce = claimer_nonce + greeter_nonce
+    # Digest size of 5 bytes so we can split it beween two 20bits SAS
+    combined_hmac = shared_secret_key.hmac(combined_nonce, digest_size=5)
+
+    hmac_as_int = int.from_bytes(combined_hmac, "big")
+    # Big endian number extracted from bits [0, 20[
+    claimer_sas = hmac_as_int % 2 ** 20
+    # Big endian number extracted from bits [20, 40[
+    greeter_sas = (hmac_as_int >> 20) % 2 ** 20
+
+    return claimer_sas, greeter_sas
+
+
+def generate_sas_code_candidates(valid_sas: int, size: int = 3) -> List[int]:
+    candidates = {valid_sas}
+    while len(candidates) < size:
+        candidates.add(randint(0, 2 ** 20 - 1))
```

Comment:
Would it make sense to generate candidates that always differ from `valid_sas`?

For instance we could make sure that the first 5 bits are always different from that `valid_sas` so that the candidates all start with a different character.

Reply:
> Would it make sense to generate candidates that always differ from valid_sas?

`candidates` is a set, so we are guaranteed it's the case

> For instance we could make sure that the first 5 bits are always different from that valid_sas so that the candidates all start with a different character.

I don't think it's a good idea given the user is supposed to check the entire code so similar starting character shouldn't be an issue (and advanced user noticing a pattern and figuring out he only have to check for the first character is a security concern !)

Reply:
> `candidates` is a set, so we are guaranteed it's the case

Oh, right!

> the user is supposed to check the entire code so similar starting character shouldn't be an issue
> (and advanced user noticing a pattern and figuring out he only have to check for the first character is a security concern !)

Reply:
> `candidates` is a set, so we are guaranteed it's the case

Oh, right!

> the user is supposed to check the entire code so similar starting character shouldn't be an issue

Fair point.

> (and advanced user noticing a pattern and figuring out he only have to check for the first character is a security concern !)

I really don't find it realistic nor a security concern.